src/client-side-encryption/state_machine.ts

```diff
@@ -360,35 +355,44 @@ export class StateMachine {
         const providerTlsOptions = tlsOptions[kmsProvider];
         if (providerTlsOptions) {
           const error = this.validateTlsOptions(kmsProvider, providerTlsOptions);
-          if (error) reject(error);
+          if (error) throw error;
           try {
             await this.setTlsOptions(providerTlsOptions, options);
-          } catch (error) {
-            return onerror(error);
+          } catch (err) {
+            throw onerror(err);
           }
         }
       }
+
       socket = tls.connect(options, () => {
         socket.write(message);
       });
 
-      socket.once('timeout', ontimeout);
-      socket.once('error', onerror);
+      const { promise: onSocketDataFullyRead, reject, resolve } = promiseWithResolvers();
+      socket
+        .once('timeout', () => reject(ontimeout()))
+        .once('error', err => reject(onerror(err)))
+        .once('close', () => reject(onclose()))
+        .on('data', data => {
+          buffer.append(data);
+          while (request.bytesNeeded > 0 && buffer.length) {
+            const bytesNeeded = Math.min(request.bytesNeeded, buffer.length);
+            request.addResponse(buffer.read(bytesNeeded));
+          }
 
-      socket.on('data', data => {
-        buffer.append(data);
-        while (request.bytesNeeded > 0 && buffer.length) {
-          const bytesNeeded = Math.min(request.bytesNeeded, buffer.length);
-          request.addResponse(buffer.read(bytesNeeded));
-        }
+          if (request.bytesNeeded <= 0) {
+            // There's no need for any more activity on this socket at this point.
+            destroySockets();
+            resolve(undefined);
```

Comment:
```suggestion
            resolve();
```

`promiseWithResolvers<void>();` in order to make resolve a "no argument" function you have to parameterize the Promise with void.